File: mirea/cources/2026-2027/aii-july/03/Part-01-Ordered-set/examples/example-03.py
from ordered_set_treap import OrderedSet


def print_set(name, s):
    print(f"{name} = {{{','.join(map(str, s))}}}")


s = OrderedSet()
for x in [6, 1, 1, 7, 7, 5, 4, 4, 2]:
    s.insert(x)

print_set("s", s)
print("size =", s.size())
print("empty =", s.empty())
print("count(7) =", s.count(7))
print("count(3) =", s.count(3))
print("find(6) =", s.find(6))
print("find(3) =", s.find(3))
print("lower_bound(3) =", s.lower_bound(3))
print("upper_bound(5) =", s.upper_bound(5))
print("find_by_order(0) =", s.find_by_order(0))
print("find_by_order(3) =", s.find_by_order(3))
print("s[4] =", s[4])
print("order_of_key(5) =", s.order_of_key(5))
print("erase(4) =", s.erase(4))
print_set("s", s)
print("erase(4) =", s.erase(4))
print_set("s", s)
print("insert(3) =", s.insert(3))
print("insert(3) =", s.insert(3))
print_set("s", s)

File: mirea/cources/2026-2027/aii-july/03/Part-02-Bounds-and-multiset/examples/example-03.cpp
#include <ext/pb_ds/assoc_container.hpp>
#include <ext/pb_ds/tree_policy.hpp>
#include <functional>
#include <iostream>
using namespace std;
using namespace __gnu_pbds;

template<typename T, typename Comp = std::less<T>>
using ordered_set = tree<T, null_type, Comp, rb_tree_tag, tree_order_statistics_node_update>;

void print(const string &name, const auto &a) {
    cout << name << " = {";
    int i = 0;
    for (int x : a) {
        if (i > 0) cout << ',';
        cout << x;
        i++;
    }
    cout << "}" << '\n';
}

int main() {
    ordered_set<int> s;

    // Вставка.
    s.insert(10);
    s.insert(3);
    s.insert(7);
    s.insert(15);
    s.insert(7);
    print("after insert", s);

    // find_by_order(i) возвращает итератор на i-й элемент.
    for (int i = 0; i < (int)s.size(); i++) {
        cout << "s[" << i << "] = " << *s.find_by_order(i) << '\n';
    }

    // Удаление.
    s.erase(10);
    print("after erase(10)", s);

    // Для int количество элементов на отрезке [L, R]
    // равно order_of_key(R + 1) - order_of_key(L).
    int L = 4;
    int R = 15;
    int cnt = s.order_of_key(R + 1) - s.order_of_key(L);
    cout << "count in [4, 15] = " << cnt << '\n';
}
